src/plots/barPlot.ts

```diff
@@ -525,69 +534,126 @@ export class Bar<X, Y> extends XYPlot<X, Y> {
         labelOrigin.x += containerWidth / 2 - measurement.width / 2;
 
         const barY = attrToProjector["y"](d, i, dataset);
+
+        // height of the bar in the viewport
         let effectiveBarHeight = barHeight;
         if (barY + barHeight > this.height()) {
           effectiveBarHeight = this.height() - barY;
         } else if (barY < 0) {
           effectiveBarHeight = barY + barHeight;
         }
         const offset = Bar._LABEL_PADDING;
-        showLabelOnBar = measurement.height + 2 * offset <= effectiveBarHeight;
+        showLabelOnBar = (measurement.height + 2 * offset <= effectiveBarHeight)
+          && (this._labelsPosition !== LabelsPosition.outside);
+
+        const placeLabel = (position: "top" | "middle" | "bottom" | "outside-top" | "outside-bottom") => {
+          switch (position) {
```

Comment:
DRY the vertical and horizontal cases please, and move the label drawing logic into its own method imo